Add RFC contribuiting guidelines and template (#442)
Farfetch/kafkaflow: rfcs/0000-rfc-template.md

---
- Feature Name: (fill me in with a unique ident, `my_feature_name`)
- Start Date: (fill me in with today's date, YYYY-MM-DD)

> - RFC PR(s): (add the reference to the PR where this RFC was proposed and discussed, can only be added after the PR has been created)

Reviewer: Question: Isn't it a better workflow to have discussions as part of the Issues and then, implementation PRs can link back to either the issue or related PRs?

Author: Ok, it's a valid point. I think we need to define and document this, from what I see we have 2 possible workflows:

1. The issue is opened, and the RFC is created after that all the discussions happen on the RFC PR.
2. All the discussion happens on the issue, the RFC is to document the final decision.

The process we follow (in our use case) is to centralize the discussions on the RFC PR, because when you are creating the RFC is when you will have more context (usually is when you have a POC also to assist the creation of the RFC).
But we need to think about which option better suits the community. Also tagging others to leave their feedback (@filipeesch  @brmagadutra @ailtonguitar @simaoribeiro @erik-catalao @ruiqbarbosa @JoaoRodriguesGithub )

Reviewer: The problem that I see with having the discussion on the RFC is that it's not a natural flow. 
Asking someone to Fork, Commit and submit a PR, doesn't seem a good async process to leave a simple question.

Reviewer 2: I don't see any problem with the RFC process, for me the document is easier to discuss (it is the same flow as a PR review), easier to find (since all the design docs will be at the same place), and we keep it simple where the issues are a place to ask for feature or to report any bug/problem.

My main concern with the issues being the center of the discussion is regarding history, comments and discussions can be removed, the issue could be huge, and navigating through it and finding the necessary information will be hard. How would we check regarding the evolution and changes in the issue (comparing what was pointed out to what was changed)? How this process on issues can interfere with the metrics for the project (PR time to be merged, time to first respond to an issue, time to close an issue, etc.)?

Reviewer: Good point regarding the history. However, having to create a PR to comment isn't an acceptable flow. 
I think we can find inspiration here where they do the discussion on the Issue and then open questions might be transposed into the file.
- https://github.com/vuejs/rfcs
-  https://github.com/dotnet/csharplang

Reviewer 2: We can use the [discussions](https://github.com/Farfetch/kafkaflow/discussions) as the base and afterward transport it to a markdown document. Seems like a good tradeoff, we will still have the document with the proposal and the discussions will be documented beforehand.

Author: @gsferreira I think that the idea is not to open a PR just to ask a question, there could still be discussion on the issue level about the reported issue (further clarification, etc...).

On the RFC PR the discussion should be around what is being proposed in the RFC (Other approaches that the RFC author might not have considered for example.)

My concern with using discussions is that having the issues, the RFC PR and the discussions, could spread topics and make the information harder to find (for us and for others).

Author: Can we follow up on this? The original question was related to where the discussion would occur, within the RFC or in the issue.

I think both contexts are well-defined enough to be clear that discussion within the issue is related to the issue reported, and discussions within the RFC PR are related to what is being proposed within the RFC. I understand that there may be some overlap since both the issue and the RFC are addressing the same subject, but honestly, I think it is a non-issue since they will be cross-referenced.

Reviewer: Sounds good to me.

- Issue(s): (reference issue related with RFC proposal)
---

# RFC Name

## Summary

[summary]: #summary

This is a one paragraph explanation of the issue and the proposal.

## Motivation

[motivation]: #motivation

This section details why we are doing something, what use cases it supports, and what we expect the outcome to be.

## Guide Implementation

[guide-level-explanation]: #guide-level-explanation

> This is a high level overview of the process. In this section, explain the proposal like you were teaching it to a new comunity member. That means you're probably including:

Reviewer: ```suggestion
This is a high-level overview of the process. In this section, explain the proposal like you were teaching it to a new community member. That means you're probably including:
```

Author: Applied suggestion.

* Introduction of new named concepts
* Explain things in terms of examples

> * Explain the impact and how you want developers to approach the feature and its goals

Reviewer: ```suggestion
* Explain the impact and how the feature and its goals can be approached at implementation.
```

Author: Suggestion applied.

* It's interaction with other features
* It is reasonably clear how this would be implemented
* Corner cases are addressed through example
You can (and should) expand on the examples you used in the Guide Implementation and provide additional context.

## Drawbacks

[drawbacks]: #drawbacks

You should explain why we _wouldn't_ do this. There is a cost associated with these changes, and while there wouldn't be an RFC if the benefits didn't outweigh the drawbacks, your readers might not know all of the tradeoffs being made.

## Rationale / Alternatives

[rationale-and-alternatives]: #rationale-and-alternatives

You need to answer why this design is the right design. To do so, you need to understand all of the possible alternatives. You should list alternate designs and ideas and why they weren't considered for the solution. This ensures due diligence has been done on the proposal.

## Unresolved Questions

[unresolved-questions]: #unresolved-questions

You should gather any open questions and either list them in the document (folks often use markdown quoting) or add them to this section at the end of the document. If you want to mark unresolved options inline

> **Open Question:** this is an open question

Reviewer: This part is not clear to me. 
There's a section for questions or do I simply add a bold "Open Question:" where I want?
How does that work with collaboration through GIT? Shouldn't it be a typical pull request review workflow?

Author: This is during the RFC text to create a point of discussion and is usually for the RFC author to ask the reviewers for something.

After the questions are discussed in the PR and an agreement is reached, the RFC author can remove that question from the RFC, and add conclusions to the document if they are necessary.

Reviewer: It's related to the discussion here: https://github.com/Farfetch/kafkaflow/pull/442#discussion_r1343748551
